### DISCUSSION TITLE
`ReferenceError: performance is not defined` when deploying to Vercel

### DISCUSSION THREAD

**Discussion by EvHaus** [vercel/next.js#36209] [Help] [answered]
Hey Vercel team,

Sorry to bother you but I'm stuck on a really strange issue. In my Next.js app I'm using `@tanstack/react-table` version `8.0.0-alpha.26`. It works fine in local dev mode (`yarn dev`) and builds work fine when I generate them via `yarn build`, but when deploying to Vercel the builds on Vercel fail with:

```
ReferenceError: performance is not defined
--
23:27:09.806 | at Object.getHeaderGroups (/vercel/path0/node_modules/@tanstack/table-core/build/cjs/utils.js:64:33)
23:27:09.807 | at DataTable (/vercel/path0/.next/server/pages/reports.js:749:37)
23:27:09.807 | at Hc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:64:191)
23:27:09.807 | at Kc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:66:253)
23:27:09.807 | at Z (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:71:89)
23:27:09.808 | at Lc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:73:98)
23:27:09.808 | at Nc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:72:306)
23:27:09.808 | at Z (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:71:217)
23:27:09.811 | at Lc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:73:98)
23:27:09.811 | at Kc (/vercel/path0/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:67:131)
```

I can't figure out why this would be the case. The code on that line should not be getting executed at all. And I don't get why `yarn build` works on my machine but not on Vercel.

I tracked it down to the fact that my local environment is using a different version of Node. Node 16+ supports the `performance` API but Node 14 doesn't. This is probably what causes things to break in Vercel. Is there any way to get around this? Can I force Vercel to use Node 16?

I also logged this issue with `react-table` [here](https://github.com/TanStack/react-table/issues/3830#issuecomment-1101012448).

**Comment by icyJoseph** [comment]
I guess the page using `react-table` is being pre-rendered? 

Locally change your node version to 14, I can recommend [fnm](https://github.com/Schniz/fnm), and then:

- Use `dynamic` on the component using react-table with `{ssr: false}`, see [here](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr).
  - This would delay the execution of this code to the client, where performance is available.
- Use a client side set flag:
```ts
const [show, setShow] = useState(false);

useEffect(()=>{setShow(true)},[]);

return show && <ComponentThatUsesTable />
```

With these, or any other creative way you can think of, you could make it work at least. It does seem to be like `react-table` needs to provide `ssr` support. Unguarded access into browser/DOM API's causes this kind of issues.

---

You could make a PR to `react-table` and change to something like:

```js
if (opts.key && opts.debug) depTime = isServer() ?  0 : performance.now()
```

or

```js
if (!isServer() && opts.key && opts.debug) depTime = performance.now()
```

It depends on what's `depTime` used for of course. `const isServer = typeof window === 'undefined'` should do the trick.

**Comment by EvHaus** [comment, marked as answer]
Found [a quick hack](https://github.com/TanStack/react-table/issues/3830#issuecomment-1101012448) to get around this for now.
